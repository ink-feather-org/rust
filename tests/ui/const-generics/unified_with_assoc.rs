#![feature(generic_const_exprs)]
#![feature(impl_unified_exhaustive_const_traits)]
#![allow(incomplete_features)]

pub trait WithAssoc {
  const VAL: usize;
  type Item;
}

pub struct Demo<const N: bool>;

impl WithAssoc for Demo<true> {
  const VAL: usize = 3;
  type Item = usize;
}
impl WithAssoc for Demo<false> {
  const VAL: usize = 5;
  type Item = String;
}
fn demo_func<const N: usize>() where Demo<{ N > 5 }>: WithAssoc,
  [();Demo::<{N>5}>::VAL]:,
  //~^ ERROR cycle detected
  {
  let _ = [0; Demo::<{N>5}>::VAL];
}

fn main() {}
