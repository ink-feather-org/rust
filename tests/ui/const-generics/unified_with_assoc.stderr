error[E0391]: cycle detected when building an abstract representation for `demo_func::{constant#1}`
  --> $DIR/unified_with_assoc.rs:21:7
   |
LL |   [();Demo::<{N>5}>::VAL]:,
   |       ^^^^^^^^^^^^^^^^^^
   |
note: ...which requires building THIR for `demo_func::{constant#1}`...
  --> $DIR/unified_with_assoc.rs:21:7
   |
LL |   [();Demo::<{N>5}>::VAL]:,
   |       ^^^^^^^^^^^^^^^^^^
note: ...which requires type-checking `demo_func::{constant#1}`...
  --> $DIR/unified_with_assoc.rs:21:7
   |
LL |   [();Demo::<{N>5}>::VAL]:,
   |       ^^^^^^^^^^^^^^^^^^
   = note: ...which again requires building an abstract representation for `demo_func::{constant#1}`, completing the cycle
note: cycle used when checking that `demo_func` is well-formed
  --> $DIR/unified_with_assoc.rs:20:1
   |
LL | / fn demo_func<const N: usize>() where Demo<{ N > 5 }>: WithAssoc,
LL | |   [();Demo::<{N>5}>::VAL]:,
   | |___________________________^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0391`.
